### DISCUSSION TITLE
Error: Failed finding Lua header files. You may need to install them or configure LUA_INCDIR. - Windows 10 with Lua distribution

### DISCUSSION THREAD

**Discussion by Cussa** [keplerproject/luarocks#1282] [Q&A]
Hi,

I am using the distributed version of Lua 5.3. Based on the documentation, I could take the binary and use the `luarocks`.

So, I have a folder `D:\lua\dist` where I have both the `lua53` binaries and the `luarocks`. This path is on the PATH variable too.

However, when I try to install something, I'm receiving the following error:

```
D:\lua\dist>luarocks install busted
Installing https://luarocks.org/busted-2.0.0-1.rockspec

Error: Failed finding Lua header files. You may need to install them or configure LUA_INCDIR.
```

When I ran the `luarocks`, this is the configuration that it presents:

```
Configuration:
   Lua:
      Version    : 5.3
      Interpreter: D:\lua\dist/luarocks.exe (ok)
      LUA_DIR    : D:\lua\dist (ok)
      LUA_BINDIR : D:\lua\dist (ok)
      LUA_INCDIR : D:\lua\dist (not found)
                   ****************************************
                   Use the command

                      luarocks config variables.LUA_INCDIR <dir>

                   to fix the location
                   ****************************************
      LUA_LIBDIR : D:\lua\dist (ok)
```

Can someone help me to figure out what is happening?

Thanks in advance.

**Comment by daurnimator** [comment]
@Cussa luarocks is trying to find the lua header files (i.e. lua.h, lauxlib.h, lualib.h, etc). You need to tell it the folder to find them in.

**Reply by Cussa** [reply]
I am using the binary version of lua, not the code itself. Based on the documentation, I should not need the source code from lua to do that. Or do I got it wrong?

**Reply by daurnimator** [reply]
luarocks needs the lua headers to compile lua modules written in C.

**Reply by ramkumarkb** [reply]
@daurnimator - I am facing the same issue, even though I have all the following header files in my `include` directory -

`luaxlib.h, lua.h, lua.hpp, luaconf.h, lualib.h`

Am i missing any files or is there some other issue? Below is the output when i run the `luarocks-admin` command - 

```
Configuration:
   Lua:
      Version    : 5.4
      Interpreter: C:\Users\mickeymouse\dev\gcc-lua-install\lua\bin (ok)
      LUA_DIR    : C:\Users\mickeymouse\dev\gcc-lua-install\lua\bin (ok)
      LUA_BINDIR : C:\Users\mickeymouse\dev\gcc-lua-install\lua\bin (ok)
      LUA_INCDIR :  (not found)
                   ****************************************
                   Use the command

                      luarocks config variables.LUA_INCDIR <dir>

                   to fix the location
                   ****************************************
      LUA_LIBDIR : C:\Users\mickeymouse\dev\gcc-lua-install\lua\bin (ok)

   Configuration files:
      System  : C:/Program Files/luarocks/config-5.4.lua (not found)
      User    : C:/Users/mickeymouse/AppData/Roaming/luarocks/config-5.4.lua (ok)

   Rocks trees in use:
      C:/Users/mickeymouse/AppData/Roaming/luarocks/5.4
```

And here are the details of the `config-5.4.lua` - 

```
rocks_trees = { "C:/Users/mickeymouse/AppData/Roaming/luarocks/5.4" }
variables = {
   LV = "5.4",
   LUA = "C:\\Users\\mickeymouse\\dev\\gcc-lua-install\\lua\\bin",
   LUA_INCDIR = 'C:/Users/mickeymouse/dev/gcc-lua-install/lua/include',
   CC = 'gcc',
   LD = 'gcc',
}
verbose = false
```
I tried everything - `\\` and `/` etc and still no avail - same error...

**Reply by ramkumarkb** [reply]
After some trial-n-error, what works for me is this - 
```
luarocks-admin --lua-dir C:\Users\mickeymouse\dev\gcc-lua-install\lua\bin LUA_INCDIR='C:/Users/mickeymouse/dev/gcc-lua-install/lua/include'
```
Somehow it does not get the LUA_INCDIR from the config file, only from the command line argument.

**Reply by Cussa** [reply]
@ramkumarkb if I got correct, you have the full repo from Lua on your machine. Is that right?

**Reply by ramkumarkb** [reply]
@Cussa - Not sure whether I understood your question clearly. On my Windows-10 environment, I have installation of Lua (compiled from source using the [TDM-GCC](https://jmeubank.github.io/tdm-gcc/) and then binary installation of Lua-Rocks [as per the instruction when Lua is already installed](https://github.com/luarocks/luarocks/wiki/Installation-instructions-for-Windows). Did I answer your question?

**Reply by Cussa** [reply]
HI @ramkumarkb. It seems that my problem was that I took the Binary version, while the Luarocks uses the "uncompiled" version. I "fixed" the issue removing everything from my machine and using a docker container 😅

**Reply by MAW016** [reply]
I solved this by installing lua-devel package on my linux machine. On windows you can try this: https://github.com/rjpcomputing/luaforwindows

**Reply by victorwestmann** [reply]
This command helped me change my error message! It seems to have worked to an extent:
```luarocks-admin --lua-dir C:\lua LUA_INCDIR='C:/lua/include'```

Now I'm just getting the error messages saying:  
```Configuration:
   Lua:
      Version    : 5.4
      Interpreter: C:\lua/lua54.exe (ok)
      LUA_DIR    : C:\lua (ok)
      LUA_BINDIR : C:\lua (ok)
      LUA_INCDIR : 'C:/lua/include' (ok)
      LUA_LIBDIR : C:\lua (ok)

   Configuration files:
      System  : C:/Program Files/luarocks/config-5.4.lua (not found)
      User    : C:/Users/victo/AppData/Roaming/luarocks/config-5.4.lua (not
      found)

   Rocks trees in use:
      C:\Users\victo\AppData\Roaming/luarocks ("user")

**Reply by The-oo** [reply]
I resolve this issue by install Lua dev version 

`sudo apt install liblua5.3-dev`

from this link:  https://stackoverflow.com/questions/54813206/error-failed-finding-lua-library-you-may-need-to-configure-lua-libdir

**Reply by badiku** [reply]
terrible.   the header files are right there in the specified folder, but luarocks  insist that they are not there.

**Comment by geomorillo** [comment]
please provide an easy to install package

**Reply by Zyplonox** [reply]
on debian 11 `apt install liblua5.2-0 liblua5.2-dev`

**Reply by stokito** [reply]
For the Prosody 13 you need to install `sudo apt install liblua5.4-dev`
